graphql/query_resolver.go

```diff
@@ -963,6 +963,65 @@ func (r *queryResolver) TaskNamesForBuildVariant(ctx context.Context, projectIde
 	return buildVariantTasks, nil
 }
 
+// Waterfall is the resolver for the waterfall field.
+func (r *queryResolver) Waterfall(ctx context.Context, options WaterfallOptions) (*Waterfall, error) {
+	projectId, err := model.GetIdForProject(options.ProjectIdentifier)
+	if err != nil {
+		return nil, ResourceNotFound.Send(ctx, fmt.Sprintf("finding project with id: '%s'", options.ProjectIdentifier))
+	}
+	limit := model.DefaultWaterfallVersionLimit
+	if limitOpt := utility.FromIntPtr(options.Limit); limitOpt != 0 {
+		if limitOpt > model.MaxWaterfallVersionLimit {
+			return nil, InputValidationError.Send(ctx, fmt.Sprintf("limit exceeds max limit of %d", model.MaxWaterfallVersionLimit))
+		}
+
+		limit = limitOpt
+	}
+	requesters := options.Requesters
+	if len(requesters) == 0 {
+		requesters = evergreen.SystemVersionRequesterTypes
+	}
+
+	opts := model.WaterfallOptions{
+		Limit:      limit,
+		Requesters: requesters,
+	}
+
+	versions, err := model.GetWaterfallVersions(ctx, projectId, opts)
+	if err != nil {
+		return nil, InternalServerError.Send(ctx, fmt.Sprintf("getting waterfall versions: %s", err.Error()))
+	}
+
+	// TODO DEVPROD-10179: Add check to ensure each version has tasks that match filter...
+	// Something like this: https://github.com/evergreen-ci/evergreen/blob/bf8f12ec2eefe61f0cf9bcc594924c7be8f91d1b/graphql/query_resolver.go#L869-L938
+	// We should also be able to add a check to the version's build_variant_status field in case of a build variant filter (TODO DEVPROD-10178).
+	// For now, without filters it is guaranteed that `limit` matching versions have been returned.
+
+	// TODO DEVPROD-10177: Implementing pagination will also allow us to fetch more versions if GetWaterfallVersions does not return enough *activated* versions
+
+	buildVariants, err := model.GetWaterfallBuildVariants(ctx, versions)
```

Comment:
Yeah I definitely plan to look into this! I think it will make sense to include once DEVPROD-10177 is done since pagination is a relevant prerequisite to filtering.

Also wanted to clarify what you meant by "if the user doesn't request it"; do you mean if no versions match their filter inputs?

Reply:
Sorry, I meant if the client doesn't request it! But that sgtm!